### PR TITLE
[weekly-r290] `MemPostings.PostingsForLabelMatching()`: don't hold the mutex while matching

### PR DESCRIPTION
**Pull request by colega**
Backporting https://github.com/prometheus/prometheus/pull/14286 to the `weekly-r290` branch here.

### CODE REVIEW

File: tsdb/index/postings_test.go [hunk 1]
```diff
@@ -999,117 +1001,6 @@ func TestMemPostings_Delete(t *testing.T) {
 	require.Empty(t, expanded, "expected empty postings, got %v", expanded)
 }
 
-func TestPostingsCloner(t *testing.T) {
```
Comment (pracucci) [inline tsdb/index/postings_test.go:1002]: Why was this deleted?

Comment (jesusvazquez) [inline tsdb/index/postings_test.go:1002]: Why are you removing this test?

Comment (colega) [reply tsdb/index/postings_test.go]: Recovered it rebasing again and force-pushing.

Comment (colega) [reply tsdb/index/postings_test.go]: Bad rebase, sorry, recovered it now.
